### ISSUE TITLE
Support FSDP in PyTorch

### ISSUE THREAD

**Issue by priyakasimbeg** [mlcommons/algorithmic-efficiency#796]
It is useful to shard optimizer state across devices (to save significant memory). This reflects current practice. We want to support it.
* We want to switch from no sharding to naive model parameter sharding in both framworks.
* We will forbid (in the rules) any hacks that change the model parallelization strategy and have workload-default sharding. 
* Allow submitters to opt-out of it on a per-workload basis.

**Comment by priyakasimbeg** [comment]
From meeting minutes from @Hao-Jun Michael Shi: Challenge is ensuring that JAX and PyTorch are equivalent. PyTorch should be doable by changing the DDP wrapper to the FSDP wrapper.

**Comment by IFFranciscoME** [comment]
_...For the sake of self-referencing notes and approaches._

For the pytorch case, there are two ways for doing this: 

- [Auto-wrapping](https://pytorch.org/blog/introducing-pytorch-fully-sharded-data-parallel-api/) : Which is the case of changing the DDP to FSDP. 
- [Manual-wrapping](https://pytorch.org/docs/stable/fsdp.html) : To explore complex sharding strategies.

For the Jax case, which is the one I am less familiar with:

- [Manual-parallelism](https://jax.readthedocs.io/en/latest/notebooks/shard_map.html) : A single-program multiple-data (SPMD) multi-device parallelism API to map a function over shards of data.
- [Scalax package](https://github.com/young-geng/scalax) : An external package with to write a model and training code for a single GPU/TPU, and rely on scalax to automatically scale it up to hundreds of GPUs/TPUs.

**Comment by davidtweedle** [comment]
Hi,
I have been working on this.
- The code I have is running on cifar (on kaggle), and it seems to be fine.
- The wrapping I used is by size, but we would need to make it equivalent to the JAX code.
- I am getting some errors when I try to save the model checkpoints. This may have to do with the torch version, I am not sure.
You can see the branch I am working from here:
https://github.com/davidtweedle/algorithmic-efficiency/tree/fsdp_cifar

**Comment by davidtweedle** [comment]
Hi,
I hope it is appropriate to give a quick update on what could be going on here.
When the batch norm is updated during the training step, "module.apply" is called to update the batch norm.
This is called from the FSDP wrapper of the module which asserts that the training state must be "IDLE".
But calling apply from the FSDP wrapper means that the FSDP wrapper wants to all gather the different parameters, which is not necessary because all we want to do is tell the batch norm layers to keep track of the running stats.
So hopefully it is possible to apply "update_batch_norm_fn" without calling module.apply from the FSDP wrapper.

**Comment by priyakasimbeg** [comment]
Won't fix per discussion in the WG meetings.
